Add AboutApplicatives koan

t-rutten opened the pull request:
Here's a brief demonstration of applicatives. I'm happy to include additional patterns reviewers may suggest.

@7h3kk1d was very helpful in pulling this together.

t-rutten committed:
Intitial applicatives
zip embedded values with embedded functions

t-rutten committed:
Lazy applicatives

Co-authored-by: Alexander Bandukwala <7h3kk1d@gmail.com>

t-rutten committed:
Function is applicative

Co-authored-by: Alexander Bandukwala <7h3kk1d@gmail.com>

t-rutten committed:
Parallel computation in applicatives

Co-authored-by: Alexander Bandukwala <7h3kk1d@gmail.com>

t-rutten committed:
Omit assertion values, cleanup

7h3kk1d commented on src/test/java/com/jnape/palatable/lambdakoans/AboutApplicatives.java:
```diff
@@ -0,0 +1,157 @@
+        assertThat(just(9).zip(just(inc)), equalTo(just(10)));
+        assertThat(just(15).zip(just(inc)), equalTo(just(__)));
+        // Explicit types help the compiler
+        assertThat(Maybe.<Integer>nothing().zip(just(inc)), equalTo(__));
```
I wonder if it's useful to also have a `just(7).zip(nothing())`

7h3kk1d commented on src/test/java/com/jnape/palatable/lambdakoans/AboutApplicatives.java:
```diff
@@ -0,0 +1,157 @@
+        assertThat(right("World!").zip(right(getStringLength)), equalTo(__));
+        assertThat(Either.<String, String>left("whoops").zip(right(getStringLength)), equalTo(left(__)));
+
+        // Moving on to LambdaIterables, where "zipping" is more obvious
```
We might want to point out that `LambdaIterable` is a specific choice in terms of how to implement Zip for iterables. It could be pairwise or cross-product. Alternatively, it might be worth putting it in a separate "test" to make that explicit.

t-rutten replied:
Added a comment about that possible alternative in [5d1d6da](https://github.com/palatable/lambda-koans/pull/6/commits/5d1d6da6ea86833c9c96bd5b07835c541cfae51a).

7h3kk1d commented on src/test/java/com/jnape/palatable/lambdakoans/AboutApplicatives.java:
```diff
@@ -0,0 +1,157 @@
+
+        // Moving on to LambdaIterables, where "zipping" is more obvious
+        LambdaIterable<Integer> oneThroughThree = LambdaIterable.wrap(asList(1, 2, 3));
+        Applicative<Fn1<? super Integer, ? extends Integer>, LambdaIterable<?>> wrappedInc = LambdaIterable.wrap(asList(inc));
```
I think `LambdaIterable<Fn1<? super Integer, ? extends Integer>>` is easier to read as a type signature unless we're trying to point out that it's an applicative.

7h3kk1d commented on src/test/java/com/jnape/palatable/lambdakoans/AboutApplicatives.java:
```diff
@@ -0,0 +1,157 @@
+        LambdaIterable<Integer> zippedOnes = infiniteOnes.zip(wrappedInc);
+        assertThat(take(3, zippedOnes.unwrap()), iterates(__()));
+
+        // We might lazily get a mapping function...
+        AtomicInteger computed = new AtomicInteger(0);
+
+        Fn1<Integer, Maybe<Fn1<? super Integer, ? extends String>>> expensiveWayToGetMaybeToString = fn1((Integer n) -> {
+            for (int i = 0; i < n; i++) {
+                computed.incrementAndGet();
+            }
+            return just(Object::toString);
+        });
+
+        Lazy<Maybe<Fn1<? super Integer, ? extends String>>> lazyGetToString = lazy(() ->
+                expensiveWayToGetMaybeToString.apply(100_000_000));
+
+        // ...then apply it with lazyZip
+        Maybe<Integer> nothing = nothing();
+        Lazy<Maybe<String>> lazyNothingToString = nothing.lazyZip(lazyGetToString);
```
We should point out that `Maybe` is strict and that's what makes it different from the iterable case.

7h3kk1d commented on src/test/java/com/jnape/palatable/lambdakoans/AboutApplicatives.java:
```diff
@@ -0,0 +1,157 @@
+    }
+
+    @Test
+    public void applicativeRepresentsParallelism() throws ExecutionException, InterruptedException {
```
I wonder if we can write this test to timeout at a certain limit and the problem can be to change the executor to have enough threads to run them in parallel and make the test pass.

7h3kk1d commented:
Thanks for the contribution! From a pedagogical standpoint, I wonder if it makes sense to introduce `liftA2`, `liftA3`, etc. higher up so we're not just covering it in the context of functions being applicative. I also think `LambdaIterable` being applicative can be confusing, so I'm conflicted if it should go higher/lower.

t-rutten committed:
Various changes from feedback
- reorder functionIsApplicative
- add assertion
- change type signature
- add timeout for applicativeRepresentsParallelism
- add comments

t-rutten committed:
Comment about Cartesian product vs pairwise
